src/Voron/GlobalPrefetchingBehavior.cs

```diff
@@ -0,0 +1,105 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using Sparrow;
+using Sparrow.Collections;
+using Sparrow.Logging;
+using Sparrow.Utils;
+using Voron.Platform.Win32;
+
+namespace Voron.Impl
+{
+    public class GlobalPrefetchingBehavior
+    {
+        public unsafe struct PrefetchCommand
+        {
+            public void* Base;
+            public long Length;
+        }
+
+        private const string PrefetchingThreadName = "Voron Global Prefetching Thread";
+
+        internal static readonly Lazy<GlobalPrefetchingBehavior> GlobalPrefetcher = new Lazy<GlobalPrefetchingBehavior>(() =>
+        {
+            var prefetcher = new GlobalPrefetchingBehavior();
+            var thread = new Thread(prefetcher.VoronPrefetcher)
+            {
+                IsBackground = true,
+                Name = PrefetchingThreadName
+            };
+            thread.Start();
+            return prefetcher;
+        });
+
+        private readonly Logger _log = LoggingSource.Instance.GetLogger<GlobalPrefetchingBehavior>("Global Prefetcher");
+         
+        public readonly SingleConsumerRingBuffer<PrefetchCommand> CommandQueue = new SingleConsumerRingBuffer<PrefetchCommand>(128);
+
+        private unsafe void VoronPrefetcher()
+        {
+            NativeMemory.EnsureRegistered();
+
+            // We want this to always run, even if we dispose / create new storage env, this is 
+            // static for the life time of the process, and environments will register / unregister from
+            // it as needed
+
+
+            // We are already zeroing 512 bytes and with the default behavior can get 16MB prefetched per sys call,
+            // so that is quite enough. We expect that we'll only rarely need to do more than 32 at a go, anyway
+            const int StackSpace = 32;
+
+            var toPrefetch = stackalloc MemorySegment[StackSpace];
+            try
+            {               
+                while (true)
+                {
+                    var commands = CommandQueue.Acquire();
+                    if (commands.IsEmpty)
+                        Thread.Sleep(1000); // If no command is in the queue, we are surely not busy. 
```

Comment:
Nope. Because the caller is high frequency and we want to avoid the actual cost on the caller side. We dont care if we miss a few, in fact it goes way beyond that... we dont care if we miss a ton of them because we dont have enough space on the reservation queue.

Reply:
The caller is already protected by `ShouldPrefetch`, so I don't think we are going to see too many of that

Reply:
You will. Don't trust me, trust the profiling trace :)

Reply:
After some changes, the profiling trace suggests we can do wake up based on quantity. I tried though to get down the number of milliseconds down, with barely any change in hard page faults though.

Reply:
Can you show the numbers?

Reply:
Implemented using a BlockingCollection instead. The wait was too much on average.